Globe atmosphere + Style specification

karimnaaji opened the pull request:
## Atmosphere style specification

This PR extends the fog system to allow design of an atmosphere effect supported on both globe and mercator projections. It extends the `fog` style specification to account for three new values: `high-color`, `space-color` and `star-intensity`. The `star-intensity` allows for a parallax effect while rotating and panning the globe to increase the effect of 3d.

![Webp net-gifmaker](https://user-images.githubusercontent.com/7061573/162287984-25ceeefb-f9df-4f8a-9099-baf2e3b647da.gif)

## Changelist

- Add an `atmosphere` debug page with different preset to test the feature
- Add support for fog on g lobe with a fixed range, the fixed range value is chosen arbitrarily to give plausible rendering of a thin atmospheric layer over the globe.
- Consolidate frustum corner calculation and share it between atmosphere rendering and fog uniform upload
- Split atmosphere buffer from `GlobeSharedBuffer` so that it's not tied to globe rendering
- Remove clear color potentially using fog color, as this shader now covers the entire screen, it is unnecessary to dispatch an expensive clear
- Make sure that globe fixed fog range does not affect rendering of symbols and fog tile culling
- Make sure that heatmaps are not affected by globe fog fixed range, so that this layer can be rendered without being affected by fog opacity, as we get flickering artifacts otherwise
- Update render test baselines to account for the new changes
- Add style specification defaults using complex expressions so that `map.setFog({})` results in a nice default for the user
- Add a fake cubemap in the shader in order to create the parallax effect for stars

https://user-images.githubusercontent.com/7061573/162288266-c7804e9f-ed57-4db7-987d-65a0620c4f42.mov

## Style specification entries (including old ones):

- `color` alpha channel

![color-alpha](https://user-images.githubusercontent.com/7061573/162287696-7da34d72-4427-4b0b-b2fa-4af13ecedcbb.gif)

  - `color` rgb channels

![color](https://user-images.githubusercontent.com/7061573/162287717-8b31a077-fe86-4e32-b8df-d1562e1ab32a.gif)

- `horizon-blend`

![horizon-blend](https://user-images.githubusercontent.com/7061573/162287733-74802387-93a0-4a25-b9e7-cb2840d819e1.gif)

- `high-color` alpha channel

![sky-color-alpha](https://user-images.githubusercontent.com/7061573/162287752-7d8bb975-b610-4b1a-af19-e2b6a363a01b.gif)

- `high-color` rgb channels

![sky-color](https://user-images.githubusercontent.com/7061573/162287769-ab37e62c-1376-484a-9b4a-4043c2299bf4.gif)

- `space-color` alpha channel

![space-color-alpha](https://user-images.githubusercontent.com/7061573/162287793-75ffb3d9-8379-46e0-9a27-6328d0b573da.gif)

- `space-color` rgb channels

![space-color](https://user-images.githubusercontent.com/7061573/162287815-47059f34-1bdf-463a-91ab-a2581a90b3ab.gif)

- `star-intensity`

![star-intensity](https://user-images.githubusercontent.com/7061573/162287876-6b859efd-733d-4d93-8dcf-74348c4af464.gif)

## Follow up to address after this PR

- Reduce frequency of globe specific uniforms when updating fog uniform, these values can be 1) calculated once per frame and 2) only be done when used along with globe
- ~~Fix horizon angle line visual artifact, refer~~ https://github.com/mapbox/gl-js-team/issues/410
- Add render tests specific to globe atmosphere
- Fix fill-extrusion fog contribution during transition

## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->

 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [x] manually test the debug page
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'

mpulkki-mapbox commented on src/style/properties.js:
```diff
@@ -759,7 +760,7 @@ export class Properties<Props: Object> {
                 defaultTransitionablePropertyValue.untransitioned();
             this.defaultPossiblyEvaluatedValues[property] =
-                defaultPropertyValue.possiblyEvaluate(({}: any));
+                defaultPropertyValue.possiblyEvaluate(defaultParameters);
```
Is this related to the atmosphere code?

karimnaaji replied:
This was necessary to fix an issue that manifested itself now that we introduced default parameters in a root style specification entry. Having a dummy parameter at creation allows further dynamic evaluation of the property values.

mpulkki-mapbox replied:
okay thanks for the clarification! :+1:

mpulkki-mapbox commented on src/shaders/atmosphere.fragment.glsl:
```diff
@@ -0,0 +1,136 @@
+    return fract(p.x * p.y);
+}
+
+float stars(vec3 p, float scale, vec2 offset) {
+    vec2 uv_scale = (u_viewport / u_star_size) * scale;
+    vec3 position = vec3(p.xy * uv_scale + offset * u_viewport, p.z);
+
+    vec3 q = fract(position) - 0.5;
+    vec3 id = floor(position);
+
+    float random_visibility = step(random(id), u_star_density);
+    float circle = smoothstep(0.5 + u_star_intensity, 0.5, length(q));
+
+    return circle * random_visibility;
+}
```
This produces a really nice looking starfield! Artefacts from distributing points on the cube (instead of uniform distribution on a sphere) are pretty much indistinguishable!

mpulkki-mapbox reviewed (commented):
Looks really good! Are we supposed to have a default preset available when switching to globe view? Rest of the debug pages (e.g. terrain-debug.html) have no atmosphere set.

karimnaaji commented on src/util/primitives.js:
```diff
@@ -85,6 +85,46 @@ class Ray {
+        const horizonClipL = [-1, horizonFromTopClip, 1];
+        const horizonClipR = [1, horizonFromTopClip, 1];
+
+        const horizonL = vec3.transformMat4(horizonClipL, horizonClipL, invProj);
```
Yes, good idea. I've updated to use `vec3.lerp` in latest. I also investigate whether we actually need two values here, because I believe we could simplify for a single uniform.

karimnaaji commented on src/util/primitives.js:
```diff
@@ -85,6 +85,46 @@ class Ray {
     }
 }
 
+class FrustumCorners {
```
There's at least one place in the code that doesn't need the frustum corners, where consolidating in `Frustum` would make this extraneous: https://github.com/mapbox/mapbox-gl-js/blob/main/src/ui/camera.js#L748 so I think it's ok to be split even though it's related.

karimnaaji commented on src/style/fog.js:
```diff
@@ -63,6 +79,12 @@ class Fog extends Evented {
             return;
         }
 
+        for (const name of Object.keys(styleSpec.fog)) {
+            if (fog && fog[name] === undefined) {
+                fog[name] = styleSpec.fog[name].default;
+            }
+        }
```
Good catch, that's a bug, will fix that.

karimnaaji replied:
Fixed in latest.

karimnaaji commented on src/render/painter.js:
```diff
@@ -979,7 +986,27 @@ class Painter {
             const fogOpacity = fog.getOpacity(this.transform.pitch);
             if (fogOpacity !== 0.0) {
-                program.setFogUniformValues(context, fogUniformValues(this, fog, tileID, fogOpacity));
+                const tr = this.transform;
+                const viewMatrix = tr._camera.getWorldToCamera(tr.worldSize, 1.0);
+                const center = [tr.globeMatrix[12], tr.globeMatrix[13], tr.globeMatrix[14]];
+                const globeCenterInViewSpace = vec3.transformMat4(center, center, viewMatrix);
+                const globeRadius = tr.worldSize / 2.0 / Math.PI - 1.0;
+                const viewport = [
+                    tr.width * browser.devicePixelRatio,
+                    tr.height * browser.devicePixelRatio
+                ];
```
Yes, there are at least two improvements to make there that I mentioned in the description: 
>Reduce frequency of globe specific uniforms when updating fog uniform, these values can be 1) calculated once per frame and 2) only be done when used along with globe

because these are evaluated for every draw call, and could be evaluated once per frame instead. I'll fix that in this PR.

karimnaaji commented on src/geo/projection/globe.js:
```diff
@@ -86,6 +86,11 @@ export default class Globe extends Mercator {
     }
 
+    createFogTileMatrix(tr: Transform): Float64Array {
+        const matrix = mat4.identity(new Float64Array(16));
+        return mat4.multiply(matrix, matrix, tr.globeMatrix);
```
Simplified for `clone` in latest.

mpulkki-mapbox replied:
`mat4.clone()` returns a matrix using a Float32Array. This might have undesirable side effects?

karimnaaji replied:
Yes, good catch. Will revert from clone and simplify.

karimnaaji commented on debug/atmosphere.html:
```diff
@@ -0,0 +1,229 @@
+    // 'mapbox/outdoors-v11',
+    // 'mapbox/satellite-v9',
+    if (value === 'preset1') {
+        guiParams.enable = true;
+        guiParams.style = 'satellite-streets-v11';
+        guiParams.color = [220, 159, 159];
+        guiParams.highColor = [36, 92, 223];
+        guiParams.spaceColor = [0, 0, 0];
+        guiParams.spaceAlpha = 1.0;
+        guiParams.colorAlpha = 1.0;
+        guiParams.skyAlpha = 0.7;
+        guiParams.horizonBlend = 0.5;
+        guiParams.starIntensity = 0.15;
```
That should be fixed with latest.

mpulkki-mapbox replied:
Still seeing some issues. This is just a minor thing as all UI controls work!

https://user-images.githubusercontent.com/55925868/163238266-0d61ec61-a62d-4ad5-ae82-c37c13e6ac5d.mp4

karimnaaji replied:
Oh ok I see what you mean now, you'd like the UI to update based on the preset values! I'm not sure if that's possible with this GUI library but I'll give it a look.

karimnaaji replied:
Using `.listen()` did the trick to refresh the values: https://github.com/mapbox/mapbox-gl-js/pull/11590/commits/73566dc190d71722968eb1d015de04d3b64a47c3

karimnaaji commented:
>Are we supposed to have a default preset available when switching to globe view?

No, there's no dependency between globe-view and atmosphere at the spec level. e.g. when setting globe-view we don't automatically enable the atmosphere. To improve UX, this is something to consider fixing at a higher level in Studio/when setting default styling, so that we're not adding dependencies between different style specification entries.

>Rest of the debug pages (e.g. terrain-debug.html) have no atmosphere set.

I've updated most of the pages in latest to use the default atmosphere.

There are two follow-ups to this PR:
- Reduce frequency of operations for https://github.com/mapbox/mapbox-gl-js/pull/11590#discussion_r847794446
- There's an issue that I've noticed on fill-extrusion on globe during the transition. I'll  address that separately:

https://user-images.githubusercontent.com/7061573/163053674-7f9a8344-7f83-473e-a324-5624c79cc961.mov

Otherwise please take another look as this is ready for another review, I think I've addressed all of the other comments.

mpulkki-mapbox commented on src/shaders/atmosphere.fragment.glsl:
```diff
@@ -0,0 +1,136 @@
+    float norm_dist_from_center = length(closest_point_forward - u_globe_pos) / u_globe_radius;
+
+    if (norm_dist_from_center < 1.0) {
+        gl_FragColor = u_color;
```
early exit here?

mpulkki-mapbox commented on src/shaders/atmosphere.fragment.glsl:
```diff
@@ -0,0 +1,136 @@
+    // exponential curve
+    // [0.0, 1.0] == inside the globe, > 1.0 == outside of the globe
+    // https://www.desmos.com/calculator/l5v8lw9zby
+    float t = exp(-horizon_angle * pow(u_fadeout_range, -1.0));
```
is this an optimization to write `pow(x, -1)` instead of `1/x`?

karimnaaji replied:
Not an optimization, that's an overlook from converting the function from desmos, thanks for pointing that out I'll update it!

karimnaaji replied:
Also, I switched that function to prefer `exp` over an `sqrt` based function, so that the effect is smoother, you can notice that the `sqrt` based function that we used in `main` (in red) results in an abrupt change when crossing `y=0` due to clamping:

![Screenshot from 2022-04-13 11-29-45](https://user-images.githubusercontent.com/7061573/163246621-65cbb24a-0dbd-42d1-989c-8acae2f2e858.png)

mpulkki-mapbox replied:
yes I noticed! That is a better option :+1:

mpulkki-mapbox reviewed (approved):
LGTM!
